src/mgr/PyModuleRegistry.h

```diff
@@ -217,14 +217,9 @@ class PyModuleRegistry
     }
   }
 
-  auto get_clients() const
+  const auto& get_clients() const
   {
-    std::scoped_lock l(lock);
-    std::vector<entity_addrvec_t> v;
-    for (const auto& p : clients) {
-      v.push_back(p.second);
-    }
-    return v;
+    return clients;
```

Comment:
Note that this changes the semantics beyond just vector -> multimap.  Are callers OK with getting a reference to something that can presumably change underneath them?  Is there a more coarse lock that is held in all cases?

Reply:
The only caller of `PyModuleRegistry::get_clients()` is `MgrStandby::send_beacon()`, https://github.com/ceph/ceph/blob/main/src/mgr/MgrStandby.cc#L233

`PyModuleRegistry.clients` can be modified by  `PyModuleRegistry`'s member functions, `register_client()` and `unregister_client()` .  The two member functions do not take locks before modifying `clients`.

`ActivePyModules::register_client()` and `ActivePyModules::unregister_client()` take the  `ActivePyModules::lock` before calling `PyModuleRegistry::register_client` and `PyModuleRegistry::unregister_client()` respectively.

In the original code, `PyModuleRegistry::get_client()` takes the `PyModuleRegistry::lock` and then copies the addresses in `clients` into a new vector.   I'm not sure whether `PyModuleRegistry::lock` would have protected the `clients` from being modified underneath by `ActivePyModules::register_client()` or `ActivePyModules::unregister_client()`.

Reply:
Indeed, given that `register_client()` and `unregister_client()` don't take `PyModuleRegistry::lock`, locking the accessor appears to be pointless.  @batrick it was you who added this in df507cde8d71063d5873a42f668156e4c32e86f9, care to take a look?

Reply:
For now, I've made `PyModuleRegistry::get_clients()` return a copy of `clients` member without taking the `PyModuleRegistry::lock`.

@batrick let me know if you want me to do the following change, where I take the `PyModuleRegistry::lock` within the `PyModuleRegistry` member functions that modify and access `clients`.

```
diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index 8c070dd8cbf..5ad94e1c2a5 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -1496,8 +1496,6 @@ void ActivePyModules::cluster_log(const std::string &channel, clog_type prio,
 
 void ActivePyModules::register_client(std::string_view name, std::string addrs)
 {
-  std::lock_guard l(lock);
-
   entity_addrvec_t addrv;
   addrv.parse(addrs.data());
 
@@ -1507,8 +1505,6 @@ void ActivePyModules::register_client(std::string_view name, std::string addrs)
 
 void ActivePyModules::unregister_client(std::string_view name, std::string addrs)
 {
-  std::lock_guard l(lock);
-
   entity_addrvec_t addrv;
   addrv.parse(addrs.data());
 
diff --git a/src/mgr/PyModuleRegistry.h b/src/mgr/PyModuleRegistry.h
index 3231d7c6679..14b8339665a 100644
--- a/src/mgr/PyModuleRegistry.h
+++ b/src/mgr/PyModuleRegistry.h
@@ -204,10 +204,12 @@ public:
 
   void register_client(std::string_view name, entity_addrvec_t addrs)
   {
+    std::lock_guard l(lock);
     clients.emplace(std::string(name), std::move(addrs));
   }
   void unregister_client(std::string_view name, const entity_addrvec_t& addrs)
   {
+    std::lock_guard l(lock);
     auto itp = clients.equal_range(std::string(name));
     for (auto it = itp.first; it != itp.second; ++it) {
       if (it->second == addrs) {
@@ -219,6 +221,7 @@ public:
 
   auto get_clients() const
   {
+    std::lock_guard l(lock);
     return clients;
   }
```

Reply:
I think the patch you propose is probably right. I don't recall why I didn't include the mutex locks in the `register/unregister` methods.

Reply:
> @ajarr while I'm thinking of this now, I want to let you know that I'll be making some changes to this register client code in a fix for
> 
> https://tracker.ceph.com/issues/55606
> 
> I don't think there will be conflicts but just wanted to leave a note here so I don't forget.

Thanks @batrick . I took a quick look at the PyModuleRegistry changes.